src/linalg.jl

```diff
@@ -217,26 +217,60 @@ end
 # Norms
 _inner_eltype(v::AbstractArray) = isempty(v) ? eltype(v) : _inner_eltype(first(v))
 _inner_eltype(x::Number) = typeof(x)
-@inline _init_zero(v::StaticArray) = float(norm(zero(_inner_eltype(v))))
+@inline _init_zero(v::AbstractArray) = float(norm(zero(_inner_eltype(v))))
+
+@inline maxabs_nested(a::Number) = abs(a)
+
+function maxabs_nested(a::AbstractArray)
+    prod(size(a)) == 0 && (return _init_zero(a))
+
+    m = maxabs_nested(a[1])
+    for j = 2:prod(size(a))
+        m = max(m, maxabs_nested(a[j]))
+    end
+
+    return m
+end
+
+@generated function maxabs_nested(a::StaticArray)
+    if prod(Size(a)) == 0
+        return :(_init_zero(a))
+    end
+
+    expr = :(maxabs_nested(a[1]))
+    for j = 2:prod(Size(a))
+        expr = :(max($expr, maxabs_nested(a[$j])))
+    end
+
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds return $expr
+    end
+end
 
 @inline function LinearAlgebra.norm_sqr(v::StaticArray)
     return mapreduce(LinearAlgebra.norm_sqr, +, v; init=_init_zero(v))
 end
 
-@inline norm(a::StaticArray) = _norm(Size(a), a)
-@generated function _norm(::Size{S}, a::StaticArray) where {S}
-    if prod(S) == 0
+@generated function norm(a::StaticArray)
+    if prod(Size(a)) == 0
         return :(_init_zero(a))
     end
 
-    expr = :(LinearAlgebra.norm_sqr(a[1]))
-    for j = 2:prod(S)
-        expr = :($expr + LinearAlgebra.norm_sqr(a[$j]))
+    expr = :(LinearAlgebra.norm_sqr(a[1]/aₘ))
+    for j = 2:prod(Size(a))
+        expr = :($expr + LinearAlgebra.norm_sqr(a[$j]/aₘ))
     end
 
     return quote
         $(Expr(:meta, :inline))
-        @inbounds return sqrt($expr)
+        zero_a = _init_zero(a)
+        aₘ = maxabs_nested(a)
+        if iszero(aₘ)
+            return zero_a
+        else
+            @inbounds return aₘ * sqrt($expr)
```

Comment:
This always takes the max, and then for nonzero norm always computes with the division. How does this compare to the speed of the existing code?

The alternative strategy would be just to compute the norm, and only if that is zero or infinite, worry about scaling. More work in the worst case but hopefully most cases in real use will be neither huge nor tiny. 

Xref https://github.com/JuliaLang/julia/pull/43256

Reply:
Always taking the max and dividing by it is what is done in `LinearAlgebra` to avoid overflow and underflow; see https://github.com/JuliaLang/julia/issues/1861#ref-pullrequest-165754044.

Reply:
OK I timed things:
```julia
julia> @btime norm($(@SVector rand(10)));  # tagged version
  1.833 ns (0 allocations: 0 bytes)

julia> @btime sqrt(sum(abs2, $(@SVector rand(10))));
  1.833 ns (0 allocations: 0 bytes)

julia> @btime norm_gen($(@SVector rand(10)));  # this PR
  35.750 ns (2 allocations: 112 bytes)

julia> @btime mapreduce(abs, max, $(@SVector rand(10)));
  13.652 ns (0 allocations: 0 bytes)

julia> @btime @fastmath mapreduce(abs, max, $(@SVector rand(10)));
  2.708 ns (0 allocations: 0 bytes)
```
Here `@fastmath` will get `Inf`, `NaN` and `-0.0` wrong, but if it can be safely used, helps a lot.

Reply:
In the current Julia master, the relevant code is [here](https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/generic.jl#L510-L519).

Reply:
Yes. And is quite slow, which 43256 might fix.
```
julia> @btime LinearAlgebra.generic_norm2($(randn(10)));
  20.269 ns (0 allocations: 0 bytes)

julia> @btime sqrt(sum(abs2, $(randn(10))));
  5.208 ns (0 allocations: 0 bytes)

# longer
  
julia> @btime LinearAlgebra.generic_norm2($(randn(10^4)));
  15.750 μs (0 allocations: 0 bytes)

julia> @btime sqrt(sum(abs2, $(randn(10^4))));
  2.083 μs (0 allocations: 0 bytes)

julia> @btime norm($(randn(10^4)));  # BLAS
  12.416 μs (0 allocations: 0 bytes)
```

Reply:
But I understand your concern.  It turns out that `norm(::StaticArray)` is slower than `norm(::AbstractArray)` for the same array size:
```julia-repl
julia> v = rand(10); @btime norm($v);
  20.800 ns (0 allocations: 0 bytes)

julia> sv = @SVector rand(10); @btime norm($sv);
  28.727 ns (0 allocations: 0 bytes)
```
I will experiment with your suggestion.

Reply:
Sorry, I didn't know that the PR you Xref'ed was `julia`'s; I thought it was `StaticArrays`'s PR.  Now what you wrote makes much more sense to me.

Reply:
Just pushed a change that performs scaling only when the norm is `0` or `Inf`.  Now the benchmark is much better:
```julia-repl
julia> v = rand(10); @btime norm($v);
  21.335 ns (0 allocations: 0 bytes)

julia> sv = @SVector rand(10); @btime norm($sv);
  5.730 ns (0 allocations: 0 bytes)
```